### PR TITLE
TandemFoil: full physics feature stack (TE+wake+Cp+vortex+asinh+residual)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The senpai_transolver has multiple physics-informed feature augmentations that were developed incrementally across the #2319→#2379 lineage. Running them all together gives us the maximum-feature configuration as a complement to the no-feature baseline. If this significantly beats the baseline, we can attribute the gain to physics features collectively and then ablate individually in Round 2.

## Instructions

```bash
cd target/icml2026
python train.py \
  --dataset tandemfoil \
  --epochs 150 \
  --lr 3e-4 \
  --optimizer lion \
  --batch-size 2 \
  --model senpai_transolver \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --model-slices 96 \
  --model-mlp-ratio 4 \
  --model-dropout 0.0 \
  --cosine-t-max 150 \
  --use-ema \
  --ema-decay 0.999 \
  --ema-start-step 50 \
  --use-lookahead \
  --surface-refine \
  --surface-refine-hidden 128 \
  --surface-refine-layers 2 \
  --weight-decay 1e-4 \
  --enable-te-coord-frame \
  --enable-wake-deficit \
  --enable-wake-angle \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --cp-panel-scale 1.0 \
  --enable-vortex-panel-velocity \
  --vortex-panel-scale 0.1 \
  --enable-pressure-prior-addition \
  --asinh-pressure \
  --residual-prediction \
  --wandb-name "fern/tandem-full-physics" \
  --wandb-group "radford-tandem-round1"
```

## Baseline

No baseline established yet (Round 1). Compare against frieren/tandem-baseline-default (PR #2412) once that run completes.

- **Primary metric:** `val_primary/surface_pressure_mae` (lower is better)
- **Also report:** `legacy_noam/p_in`, `legacy_noam/p_oodc`, `legacy_noam/p_tan`, `legacy_noam/p_re`

## Results

_To be filled in by the student._

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Use this PR to run a compact parallel sensitivity sweep inside the full-physics family rather than a single run.

Launch at least these 4 variants in parallel:
- `v0`: exact full-physics command from the PR body
- `v1`: same command but remove `--enable-wake-angle`
- `v2`: same command but set `--cp-panel-scale 0.5`
- `v3`: same command but set `--vortex-panel-scale 0.05`

Use one GPU per variant with `CUDA_VISIBLE_DEVICES`. Keep the shared W&B group and suffix the run names with the variant id. If one variant clearly diverges early, kill it and replace it with `cp-panel-scale 2.0` or `vortex-panel-scale 0.2` within this same PR. I want this PR to tell us whether the full physics stack is robust or overly sensitive, not just whether one exact setting works.

**Comment by morganmcg1** [comment]
ADVISOR: Clarification on the effective run budget for this `radford` sweep.

The assignment text may show `--epochs 150`, but the live senpai harness for this run is enforcing:
- `SENPAI_MAX_EPOCHS=50`
- `SENPAI_TIMEOUT_MINUTES=30`

So in practice, your run stops at 50 epochs or 30 minutes, whichever comes first. Treat this as the true budget when deciding how many parallel variants to launch and when interpreting results.

Please report for each variant:
- epochs actually completed
- whether it hit the 30-minute timeout or the 50-epoch cap
- the best validation metric reached within that enforced budget

Optimize for best performance under this actual capped regime, not for hypothetical longer training beyond the harness limits.

**Comment by tcapelle** [comment]
STUDENT fern:

## Results — Parallel Sensitivity Sweep (Round 1)

Ran 4 variants in parallel across GPUs 0-3, all in W&B group `radford-tandem-round1`. All runs hit the 30-min timeout after 2 epochs due to the vortex panel computation bottleneck (see below).

### Primary Metrics (epoch 2, surface_pressure_mae, lower is better)

| Variant | Description | val_primary | test_primary | W&B run |
|---------|------------|:-----------:|:------------:|---------|
| **v0** | Full physics (all 11 flags) | 270.74 | 261.88 | `rnmujr2e` |
| **v1** | Full physics minus `--enable-wake-angle` | 293.44 | 283.60 | `vv3a7oju` |
| **v2** | Full physics, `--cp-panel-scale 0.5` | 285.87 | 273.40 | `qf003wiw` |
| **v3** | Full physics, `--vortex-panel-scale 0.05` | 268.10 | **260.78** | `6nwn6xck` |

### Per-Split Test Breakdown (surface_pressure_mae)

| Variant | test_eq4 (primary) | in_dist | camber_rc | camber_cruise | re_rand |
|---------|:------------------:|:-------:|:---------:|:-------------:|:-------:|
| v0 | 261.88 | 301.96 | 269.29 | 228.63 | 247.64 |
| v1 | 283.60 | 339.98 | 293.48 | 236.78 | 264.15 |
| v2 | 273.40 | 332.22 | 297.25 | 212.43 | 251.71 |
| v3 | **260.78** | **279.19** | **262.86** | 242.32 | 258.76 |

### Commands Used

```bash
# v0 (GPU 0): Full physics
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --epochs 150 --lr 3e-4 --optimizer lion --batch-size 2 --model senpai_transolver --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --model-mlp-ratio 4 --model-dropout 0.0 --cosine-t-max 50 --use-ema --ema-decay 0.999 --ema-start-step 50 --use-lookahead --surface-refine --surface-refine-hidden 128 --surface-refine-layers 2 --weight-decay 1e-4 --enable-te-coord-frame --enable-wake-deficit --enable-wake-angle --enable-cp-panel --enable-cp-panel-tandem-only --cp-panel-scale 1.0 --enable-vortex-panel-velocity --vortex-panel-scale 0.1 --enable-pressure-prior-addition --asinh-pressure --residual-prediction --wandb-name "fern/tandem-full-physics-v0" --wandb-group "radford-tandem-round1" --agent fern

# v1 (GPU 1): No wake-angle (same as v0 minus --enable-wake-angle)
# v2 (GPU 2): cp-panel-scale 0.5 (same as v0 but --cp-panel-scale 0.5)
# v3 (GPU 3): vortex-panel-scale 0.05 (same as v0 but --vortex-panel-scale 0.05)
# All v1-v3 used --cosine-t-max 150 (matching PR body)
```

### Peak Memory

Not tracked by train.py (no `torch.cuda.max_memory_allocated` logging). All 4 runs fit on single 96GB GPUs without OOM.

### What Happened

**The full physics stack is moderately robust, not fragile.** All 4 variants converged and learned meaningfully in just 2 epochs. The sensitivity pattern:

1. **Wake-angle matters most**: Removing `--enable-wake-angle` (v1) costs +21.7 on test_primary (261.88→283.60). This is the largest single-feature degradation, consistent across all test splits.

2. **Cp-panel scale is moderately sensitive**: Halving `--cp-panel-scale` from 1.0→0.5 (v2) costs +11.5 on test_primary (261.88→273.40). The effect is not symmetric across splits — camber_cruise actually improved (228.63→212.43) while in_dist and camber_rc degraded sharply.

3. **Vortex-panel scale is insensitive or slightly beneficial at lower scale**: Halving `--vortex-panel-scale` from 0.1→0.05 (v3) marginally improved test_primary by 1.1 (261.88→260.78). v3 strongly wins on in_dist (301.96→279.19) and camber_rc (269.29→262.86) but loses on camber_cruise (228.63→242.32) and re_rand (247.64→258.76).

4. **LR=0 at epoch 1**: All runs show lr=0.0 at epoch 1 and lr=3e-4 at epoch 2. This appears to be a CosineAnnealingLR interaction with the Lookahead wrapper (LR is read before `scheduler.step()` in the logging code). Despite this, all runs learn strongly — loss drops from ~1.0 to ~0.42 by epoch 2.

5. **Inf volume pressure on test_geom_camber_cruise**: All variants show `test_geom_camber_cruise/mae_vol_p = inf`, suggesting a numerical issue in specific test cases for this split.

### Critical Throughput Bottleneck

**The vortex panel computation (`compute_vortex_panel_velocity` in `core/features.py`) is a severe bottleneck.** It uses a Python for-loop over batch elements, creating [num_points, n_panels] tensors (~130K × 64) per foil per sample. This makes each training epoch ~25 minutes and limits us to only 2 epochs within the 30-minute timeout.

With only 2 epochs of training, these results represent very early-stage convergence. The relative ranking between variants may shift significantly with more training. Addressing this bottleneck (e.g., vectorizing the vortex panel loop or precomputing the features) would be the single highest-leverage improvement for the full physics stack.

### Suggested Follow-Ups

1. **Vectorize vortex panel computation** — the Python for-loop is the #1 bottleneck. Batching the [num_points, n_panels] tensor creation across samples could give 5-10x speedup and allow 20+ epochs in 30 min.
2. **Run longer with vortex_panel_scale=0.05** — v3 slightly edges v0 but the difference is within noise at 2 epochs. Longer training would clarify.
3. **Wake-angle importance suggests a dedicated ablation** — it's the most impactful single feature. Worth understanding why (is it the feature itself or the +1 input dim?).
4. **Fix the inf volume pressure bug** — investigate `test_geom_camber_cruise/mae_vol_p = inf` across all variants.
5. **Add peak GPU memory logging** to train.py for future runs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2413 because ADVISOR: Closing — full physics stack (best val=268.10) is worse than core physics subset (262.82, #2414). The vortex-panel computation has a severe Python for-loop bottleneck making each epoch ~25 min — only 2 epochs possible. **Key findings:** (1) wake-angle is the most impactful individual feature (+22.7 val MAE when removed), (2) lower vortex-panel-scale=0.05 slightly better than 0.1, (3) full stack not worth the computational cost vs core subset. Core physics + no-EMA is the right path forward (being tested in Round 3). Thank you fern..
